autopush/router/fcmv1client.py

```diff
@@ -0,0 +1,142 @@
+                # things are handled in FCM, e.g. retransmit requests.
+                # For now, catalog them as errors and provide back-pressure.
+                err = data.get("error")
+                raise RouterException("{}: {}".format(err.get("status"),
+                                                      err.get("message")))
+            if "name" in data:
+                self.success = 1
+        except (TypeError, ValueError, KeyError, AttributeError):
+            raise RouterException(
+                "Unknown error response: {}".format(content))
+        return self
+
+
+class FCMv1(object):
+    def __init__(self,
+                 project_id,
+                 service_cred_path=None,
+                 logger=None,
+                 metrics=None,
+                 **options):
+        self.project_id = project_id
+        self.endpoint = ("https://fcm.googleapis.com/v1/"
+                         "projects/{}/messages:send".format(self.project_id))
+
+        self.token = None
+        self.metrics = metrics
+        self.logger = logger or Logger()
+        self._options = options
+        if service_cred_path:
+            self.svc_cred = ServiceAccountCredentials.from_json_keyfile_name(
+                service_cred_path,
+                ["https://www.googleapis.com/auth/firebase.messaging"])
+        self._sender = treq.post
+
+    def _get_access_token(self):
+        return self.svc_cred.get_access_token()
+
+    def _build_message(self, token, notif):
+        msg = {
+            "token": token,
+            # Specify the various formats (we use android only)
+            "android": {
+                # TTL is a duration string e.g. ("60s")
+                "ttl": str(int(notif.get("ttl", 0)))+"s",
+                "data": notif.get("data_message")
+            },
+        }
+        # Wrap up the whole thing in a "message" tag.
+        return {"message": msg}
+
+    def process(self, response, payload=None):
+        if response.code == 401:
+            raise FCMAuthenticationError("Authentication Error")
+
+        result = Result(response)
+
+        d = response.text()
+        d.addCallback(result.parse_response)
+        return d
+
+    def error(self, failure):
+        if isinstance(failure.value, FCMAuthenticationError) or \
+                isinstance(failure.value, TimeoutError) or \
+                isinstance(failure.value, ConnectError):
+            raise failure.value
+        self.logger.error("FCMv1Client failure: {}".format(failure.value))
+        raise RouterException("Server error: {}".format(failure.value))
+
+    def send(self, token, payload):
+        atoken = self._get_access_token()
+        headers = {
+            'Authorization': 'Bearer {}'.format(atoken.access_token),
+            'Content-Type': 'application/json; UTF-8'
+        }
+        message = self._build_message(token, payload)
+        if 'timeout' not in self._options:
+            self._options['timeout'] = 3
```

Comment:
gcmclient also defaults to 3 but then gcm.py overwrites it with its own default of 10 (should we set this to 10 somewhere as too?)

Reply:
Well, GCM is different, and the value is only set to 10 if it's not already present, so it's more a backup than anything else. That said, I should be better about having a "global base timeout" value.